<a href="https://colab.research.google.com/github/ddodds42/DS-Unit-2-Kaggle-Challenge/blob/master/DAVID_DODDS_DS_Sprint_Challenge_6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

_Lambda School Data Science, Unit 2_
 
# Sprint Challenge: Predict Steph Curry's shots 🏀

For your Sprint Challenge, you'll use a dataset with all Steph Curry's NBA field goal attempts. (Regular season and playoff games, from October 28, 2009, through June 5, 2019.) 

You'll predict whether each shot was made, using information about the shot and the game. This is hard to predict! Try to get above 60% accuracy. The dataset was collected with the [nba_api](https://github.com/swar/nba_api) Python library.

In [0]:
%%capture
import sys

if 'google.colab' in sys.modules:
    # Install packages in Colab
    !pip install category_encoders==2.*
    !pip install pandas-profiling==2.*

In [0]:
# Read data
import pandas as pd
url = 'https://drive.google.com/uc?export=download&id=1fL7KPyxgGYfQDsuJoBWHIWwCAf-HTFpX'
df = pd.read_csv(url)

# Check data shape
assert df.shape == (13958, 20)

## 1. Begin with baselines for classification. 

>Your target to predict is `shot_made_flag`. What would your baseline accuracy be, if you guessed the majority class for every prediction?

In [250]:
df.head()

game_id  game_event_id  ...     season_type  scoremargin_before_shot
0  20900015              4  ...  Regular Season                      2.0
1  20900015             17  ...  Regular Season                      0.0
2  20900015             53  ...  Regular Season                     -4.0
3  20900015            141  ...  Regular Season                     -4.0
4  20900015            249  ...  Regular Season                      0.0

[5 rows x 20 columns]

In [251]:
df.shot_made_flag.value_counts(normalize=True)

0    0.527081
1    0.472919
Name: shot_made_flag, dtype: float64

##Baseline = 52.71%. If we predicted every shot was a miss, we would be correct 52.71% of the time.

## 2. Hold out your test set.

>Use the 2018-19 season to test. NBA seasons begin in October and end in June. You'll know you've split the data correctly when your test set has 1,709 observations.

In [252]:
df['game_date'] = pd.to_datetime(df['game_date'], infer_datetime_format=True)
df['year'] = df['game_date'].dt.year
df['month'] = df['game_date'].dt.month
df['day'] = df['game_date'].dt.day
df.head()

game_id  game_event_id    player_name  ...  year  month  day
0  20900015              4  Stephen Curry  ...  2009     10   28
1  20900015             17  Stephen Curry  ...  2009     10   28
2  20900015             53  Stephen Curry  ...  2009     10   28
3  20900015            141  Stephen Curry  ...  2009     10   28
4  20900015            249  Stephen Curry  ...  2009     10   28

[5 rows x 23 columns]

In [0]:
df['home_crt_adv'] = df['htm'] == 'GSW'

In [0]:
df['period_seconds'] = (df['minutes_remaining']*60) + df['seconds_remaining']

In [255]:
df['period'].value_counts()

3    4129
1    4093
2    2941
4    2671
5     118
6       6
Name: period, dtype: int64

In [256]:
df.iloc[3,3]

2

In [0]:
# gamesec = []

# for i in df.index:
#   if df.iloc[i,3] > 3:
#     gamesec.append(df.loc[i,'period_seconds'])
#   else:
#     gamesec.append(((4-df.loc[i,'period'])*720)+df.loc[i,'period_seconds'])

# df['game_seconds'] = gamesec
# df

In [258]:
# train = df[df['game_date'] <= ('2018-08-01')]
explore = df[df['game_date'] <= ('2018-08-01')]
train = explore[explore['game_date'] <= ('2017-08-01')]
val = explore[explore['game_date'] >= ('2017-08-01')]
test = df[df['game_date'] >= ('2018-08-01')]

# print(train.shape, test.shape)
print(train.shape, val.shape, test.shape)

(11081, 25) (1168, 25) (1709, 25)


## 3. Engineer new feature.

>Engineer at least **1** new feature, from this list, or your own idea.
>
>- **Homecourt Advantage**: Is the home team (`htm`) the Golden State Warriors (`GSW`) ?
>- **Opponent**: Who is the other team playing the Golden State Warriors?
>- **Seconds remaining in the period**: Combine minutes remaining with seconds remaining, to get the total number of seconds remaining in the period.
>- **Seconds remaining in the game**: Combine period, and seconds remaining in the period, to get the total number of seconds remaining in the game. A basketball game has 4 periods, each 12 minutes long.
>- **Made previous shot**: Was Steph Curry's previous shot successful?

    

## Engineered 5 new features before the train-test split: year, month, day, home_crt_adv, and period_seconds. Maybe isolating these temporal variables into their own feature may be useful. I'm also going to drop the columns 'player_name' and 'shot_zone_range' because of their redundance.

In [0]:
def drop_redundant(set):
  set = set.drop(columns=['player_name', 'minutes_remaining', 'seconds_remaining', 'game_date'])
  return set

In [260]:
train = drop_redundant(train)
val = drop_redundant(val)
test = drop_redundant(test)

# print(train.shape, test.shape)
print(train.shape, val.shape, test.shape)

(11081, 21) (1168, 21) (1709, 21)


In [261]:
target = 'shot_made_flag'
X_train = train.drop(columns=target)
y_train = train[target]
X_val = val.drop(columns=target)
y_val = val[target]
X_test = test.drop(columns=target)
y_test = test[target]

# X_train.shape, y_train.shape, X_test.shape, y_test.shape
X_train.shape, y_train.shape, X_val.shape, y_val.shape, X_test.shape, y_test.shape

((11081, 20), (11081,), (1168, 20), (1168,), (1709, 20), (1709,))

In [262]:
X_train.head()

game_id  game_event_id  period  ... day home_crt_adv period_seconds
0  20900015              4       1  ...  28         True            685
1  20900015             17       1  ...  28         True            571
2  20900015             53       1  ...  28         True            362
3  20900015            141       2  ...  28         True            589
4  20900015            249       2  ...  28         True            139

[5 rows x 20 columns]

## **4. Decide how to validate** your model. 

>Choose one of the following options. Any of these options are good. You are not graded on which you choose.
>
>- **Train/validate/test split: train on the 2009-10 season through 2016-17 season, validate with the 2017-18 season.** You'll know you've split the data correctly when your train set has 11,081 observations, and your validation set has 1,168 observations.
>- **Train/validate/test split: random 80/20%** train/validate split.
>- **Cross-validation** with independent test set. You may use any scikit-learn cross-validation method.

## 5. Use a scikit-learn pipeline to encode categoricals and fit a Decision Tree or Random Forest model.

In [0]:
import category_encoders as ce
from sklearn.impute import SimpleImputer
from sklearn.ensemble import RandomForestClassifier
from sklearn.pipeline import make_pipeline
from sklearn.model_selection import cross_val_score, cross_val_predict, cross_validate
from sklearn.metrics import accuracy_score

## 6.Get your model's validation accuracy

> (Multiple times if you try multiple iterations.)

In [264]:
%%time
pipeline = make_pipeline(
    ce.OrdinalEncoder(),
    SimpleImputer(strategy='median'),
    RandomForestClassifier(random_state=42, n_jobs=-1)
)

pipeline.fit(X_train, y_train)
print('RF Validation Accuracy =', pipeline.score(X_val, y_val))

RF Validation Accuracy = 0.5933219178082192
CPU times: user 3.37 s, sys: 44.6 ms, total: 3.41 s
Wall time: 1.99 s


## 7. Get your model's test accuracy

> (One time, at the end.)

In [265]:
y_pred = pipeline.predict(X_test)
print('RF Test Accuracy = ', accuracy_score(y_pred, y_test))

RF Test Accuracy =  0.6424809830310123


### Stretch Goals
- [x] Engineer 4+ new features total, either from the list above, or your own ideas.
- [ ]  Make 2+ visualizations to explore relationships between features and target.
- [ ]  Optimize 3+ hyperparameters by trying 10+ "candidates" (possible combinations of hyperparameters). You can use `RandomizedSearchCV` or do it manually.
- [ ]  Get and plot your model's feature importances.

To demonstrate mastery on your Sprint Challenge, do all the required, numbered instructions in this notebook.

To earn a score of "3", also do all the stretch goals.

You are permitted and encouraged to do as much data exploration as you want.

**1. Begin with baselines for classification.** Your target to predict is `shot_made_flag`. What is your baseline accuracy, if you guessed the majority class for every prediction?

**2. Hold out your test set.** Use the 2018-19 season to test. NBA seasons begin in October and end in June. You'll know you've split the data correctly when your test set has 1,709 observations.

**3. Engineer new feature.** Engineer at least **1** new feature, from this list, or your own idea.
- **Homecourt Advantage**: Is the home team (`htm`) the Golden State Warriors (`GSW`) ?
- **Opponent**: Who is the other team playing the Golden State Warriors?
- **Seconds remaining in the period**: Combine minutes remaining with seconds remaining, to get the total number of seconds remaining in the period.
- **Seconds remaining in the game**: Combine period, and seconds remaining in the period, to get the total number of seconds remaining in the game. A basketball game has 4 periods, each 12 minutes long.
- **Made previous shot**: Was Steph Curry's previous shot successful?

**4. Decide how to validate** your model. Choose one of the following options. Any of these options are good. You are not graded on which you choose.
- **Train/validate/test split: train on the 2009-10 season through 2016-17 season, validate with the 2017-18 season.** You'll know you've split the data correctly when your train set has 11,081 observations, and your validation set has 1,168 observations.
- **Train/validate/test split: random 80/20%** train/validate split.
- **Cross-validation** with independent test set. You may use any scikit-learn cross-validation method.

**5.** Use a scikit-learn **pipeline** to **encode categoricals** and fit a **Decision Tree** or **Random Forest** model.

**6.** Get your model's **validation accuracy.** (Multiple times if you try multiple iterations.) 

**7.** Get your model's **test accuracy.** (One time, at the end.)


**8.** Given a **confusion matrix** for a hypothetical binary classification model, **calculate accuracy, precision, and recall.**



## 8. Given a confusion matrix, calculate accuracy, precision, and recall.

Imagine this is the confusion matrix for a binary classification model. Use the confusion matrix to calculate the model's accuracy, precision, and recall.

<table>
  <tr>
    <td colspan="2" rowspan="2"></td>
    <td colspan="2">Predicted</td>
  </tr>
  <tr>
    <td>Negative</td>
    <td>Positive</td>
  </tr>
  <tr>
    <td rowspan="2">Actual</td>
    <td>Negative</td>
    <td style="border: solid">85</td>
    <td style="border: solid">58</td>
  </tr>
  <tr>
    <td>Positive</td>
    <td style="border: solid">8</td>
    <td style="border: solid"> 36</td>
  </tr>
</table>

### Calculate accuracy 

In [0]:
correct = 85+36

In [0]:
total = correct + 58 + 8

In [268]:
accuracy = correct / total
accuracy

0.6470588235294118

### Calculate precision

In [269]:
negative_prediction_precision = 85/(85+8)
negative_prediction_precision

0.9139784946236559

In [270]:
positive_prediction_precision = 36/(58+36)
positive_prediction_precision

0.3829787234042553

### Calculate recall

In [271]:
negative_prediction_recall = 85/(85+58)
negative_prediction_recall

0.5944055944055944

In [272]:
positive_prediction_recall = 36/(36+8)
positive_prediction_recall

0.8181818181818182